python/kwiver/sprokit/adapters/adapter_data_set.cxx

```diff
@@ -0,0 +1,131 @@
+ *    and/or other materials provided with the distribution.
+ *
+ *  * Neither name of Kitware, Inc. nor the names of any contributors may be used
+ *    to endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <sprokit/processes/adapters/adapter_data_set.h>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl_bind.h>
+#include <memory>
+
+// Type conversions
+#include <vital/types/image_container.h>
+#include <vital/types/detected_object_set.h>
+#include <vital/types/descriptor_set.h>
+#include <vital/types/track_set.h>
+#include <vital/types/object_track_set.h>
+#include <vital/types/feature_track_set.h>
+#include <vital/types/timestamp.h>
+#include <vital/types/geo_polygon.h>
+#include <vital/types/homography_f2f.h>
+
+using namespace pybind11;
+using namespace kwiver::adapter;
+
+PYBIND11_MODULE(adapter_data_set, m)
+{
+  enum_<adapter_data_set::data_set_type>(m, "DataSetType"
+    , "Type of data set.")
+    .value("data", adapter_data_set::data)
+    .value("end_of_input", adapter_data_set::end_of_input)
+  ;
+
+  class_< adapter_data_set, std::shared_ptr<adapter_data_set > >(m, "AdapterDataSet")
+    .def_static("create", &adapter_data_set::create
+        , (arg("type") = adapter_data_set::data_set_type::data))
+
+    .def("__iter__", [](adapter_data_set &self)
+                                      {
+                                        return make_iterator(self.cbegin(),self.cend());
+                                      }
+                        , keep_alive<0,1>())
+    .def("type", &adapter_data_set::type)
+    .def("is_end_of_data", &adapter_data_set::is_end_of_data)
+    .def("add_value", [] (adapter_data_set &self, ::sprokit::process::port_t const& port, object const& obj)
+    {
+      self.add_value<kwiver::vital::any>(port, obj);
+    })
+    .def("add_datum", &adapter_data_set::add_datum)
+
+    // The add_value function is templated.
+    // To bind the function, we must bind explicit instances of it,
+    // each with a different type.
+    // First the native C++ types
+    .def("add_int", &adapter_data_set::add_value<int>)
```

Comment:
Is it possible to template pybind11 bindings? 

Looking at this issue it may be possible: https://github.com/pybind/pybind11/issues/199 

It may be better to just explicitly create all possible values, I'm not sure.

Reply:
I believe that in that issue, they are binding explicit instances of it. I haven't seen any other way to do it.